src/ModelContextProtocol/Client/AutoDetectingClientSessionTransport.cs

```diff
@@ -0,0 +1,200 @@
+        _httpClient = httpClient;
+        _loggerFactory = loggerFactory;
+        _logger = (ILogger?)loggerFactory?.CreateLogger<AutoDetectingClientSessionTransport>() ?? NullLogger.Instance;
+        _name = endpointName;
+        
+        // Unbounded channel to prevent blocking on writes
+        _messageChannel = Channel.CreateUnbounded<JsonRpcMessage>(new UnboundedChannelOptions
+        {
+            SingleReader = true,
+            SingleWriter = false,
+        });
+    }
+
+    public ChannelReader<JsonRpcMessage> MessageReader => _messageChannel.Reader;
+
+    /// <inheritdoc/>
+    public async Task SendMessageAsync(JsonRpcMessage message, CancellationToken cancellationToken = default)
+    {
+        if (_streamableHttpTransport == null && _sseTransport == null)
+        {
+            var rpcRequest = message as JsonRpcRequest;
+            
+            // The first message must be an initialize request
+            Debug.Assert(rpcRequest != null && rpcRequest.Method == RequestMethods.Initialize, 
+                "First message must be an initialize request");
+
+            // Try StreamableHttp first
+            _streamableHttpTransport = new StreamableHttpClientSessionTransport(_options, _httpClient, _loggerFactory, _name);
+            
+            try
+            {
+                var response = await _streamableHttpTransport.SendInitialRequestAsync(message, cancellationToken).ConfigureAwait(false);
+                
+                // If the status code is not success, fall back to SSE
+                if (!response.IsSuccessStatusCode)
+                {
+                    _logger.LogDebug("Streamable HTTP transport failed for {EndpointName} with status code {StatusCode}, falling back to SSE transport",
+                        _name, response.StatusCode);
+                    
+                    await _streamableHttpTransport.DisposeAsync().ConfigureAwait(false);
+                    _streamableHttpTransport = null;
+                    
+                    await InitializeSseTransportAsync(message, cancellationToken).ConfigureAwait(false);
+                    return;
+                }
+                
+                // Process the response
+                var responseContent = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
+                if (response.Content.Headers.ContentType?.MediaType == "application/json")
+                {
+                    await ProcessMessageFromStreamableHttpAsync(responseContent, cancellationToken).ConfigureAwait(false);
+                }
+                else if (response.Content.Headers.ContentType?.MediaType == "text/event-stream")
+                {
+                    using var responseBodyStream = await response.Content.ReadAsStreamAsync(cancellationToken);
+                    await ProcessSseResponseFromStreamableHttpAsync(responseBodyStream, rpcRequest, cancellationToken).ConfigureAwait(false);
+                }
```

Comment:
Delete this. We shouldn't have to process the response. That should be completely handled by either the _streamableHttpTransport or _sseTransport.